sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -551,3 +552,30 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
       }
   }
 }
+
+/**
+ * This rule rewrites uncorrelated PredicateSubquery expressions such as Exists.
+ * The uncorrelated Exists can be evaluated using a subplan instead of a semi-join.
+ * Also, we can use `limit 1` and `select 1` after the subquery to reduce the result set.
+ * Example:
+ * exists(select b from t where t.a = 2) => exists(select 1 from t where t.a = 2 limit 1)
+ */
+object RewriteUncorrelatedSubquery extends Rule[LogicalPlan] {
+  /**
+   * Wrap the subquery with `limit 1` and `project 1`.
+   */
+  def buildSubquery(sub: LogicalPlan): LogicalPlan = {
+    Project(Seq(Alias(Literal.create(1, IntegerType), "1")()),
+      Limit(Literal.create(1, IntegerType), sub))
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Filter(condition, child) =>
+      val newCondition = condition transform {
+        case Exists(sub, children, _) if children.isEmpty =>
```

Comment:
Only deal with `Exists`? Will it process `InSubquery`?

Reply:
I propose push the outer values of 'IN' into the subquery. e.g., `3 in (select a from t)` => `3 in (select a from t where a is null or a = 3)`, but we need to be careful about the correctness.

Reply:
Is this change committed now? Or I miss it?

Reply:
I committed it before, but I revert it now...Because I'm not sure if it is a correct conversion.